### ISSUE TITLE
Question regarding the use of PCD and number of DRDBs

### ISSUE THREAD

**Issue by mtntruong** [liuzhen03/ADNet#3]
Hi, thank you for your wonderful work in HDR imaging. In the meantime, I made an attempt to implement your algorithm. Unfortunately it was not successful. I would like to have some short questions.

- I re-use the code from [EDVR Repository](https://github.com/xinntao/EDVR/blob/master/basicsr/models/archs/edvr_arch.py) without any modifications (line 376-402). I concatenate LDR images to create a tensor size of b\*t\*c\*h\*w (following line 359) for input. Is this the correct use of PCD alignment?
- How many DRDBs were used? I used three blocks (same as in the AHDRNet) but the results were not good.
- Are <img src="https://render.githubusercontent.com/render/math?math=f_2^t"> and <img src="https://render.githubusercontent.com/render/math?math=f_2^l"> the results of the same convolutional filters?

Thank you for spending time reading my question.

**Comment by liuzhen03** [comment]
Thanks for your appreciation. For your questions:
1.  We build our DCN through [DCN_v2](https://github.com/CharlesShang/DCNv2). you should insert the following code into dcn_v2.py first.

`class PCD_Align(nn.Module):

    def __init__(self, nf=64, groups=8):
        super(PCD_Align, self).__init__()
        # L3: level 3, 1/4 spatial size
        self.L3_offset_conv1 = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for diff
        self.L3_offset_conv2 = nn.Conv2d(nf, nf, 3, 1, 1, bias=True)
        self.L3_dcnpack = DCN_sep(nf, nf, 3, stride=1, padding=1, dilation=1, deformable_groups=groups)
                              # extra_offset_mask=True)
        # L2: level 2, 1/2 spatial size
        self.L2_offset_conv1 = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for diff
        self.L2_offset_conv2 = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for offset
        self.L2_offset_conv3 = nn.Conv2d(nf, nf, 3, 1, 1, bias=True)
        self.L2_dcnpack = DCN_sep(nf, nf, 3, stride=1, padding=1, dilation=1, deformable_groups=groups)
                              # extra_offset_mask=True)
        self.L2_fea_conv = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for fea
        # L1: level 1, original spatial size
        self.L1_offset_conv1 = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for diff
        self.L1_offset_conv2 = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for offset
        self.L1_offset_conv3 = nn.Conv2d(nf, nf, 3, 1, 1, bias=True)
        self.L1_dcnpack = DCN_sep(nf, nf, 3, stride=1, padding=1, dilation=1, deformable_groups=groups)
                              # extra_offset_mask=True)
        self.L1_fea_conv = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for fea
        # Cascading DCN
        self.cas_offset_conv1 = nn.Conv2d(nf * 2, nf, 3, 1, 1, bias=True)  # concat for diff
        self.cas_offset_conv2 = nn.Conv2d(nf, nf, 3, 1, 1, bias=True)

        self.cas_dcnpack = DCN_sep(nf, nf, 3, stride=1, padding=1, dilation=1, deformable_groups=groups)
                               # extra_offset_mask=True)

        self.lrelu = nn.LeakyReLU(negative_slope=0.1, inplace=True)

    def forward(self, nbr_fea_l, ref_fea_l):
        '''align other neighboring frames to the reference frame in the feature level
        nbr_fea_l, ref_fea_l: [L1, L2, L3], each with [B,C,H,W] features
        '''
        # L3
        L3_offset = torch.cat([nbr_fea_l[2], ref_fea_l[2]], dim=1)
        L3_offset = self.lrelu(self.L3_offset_conv1(L3_offset))
        L3_offset = self.lrelu(self.L3_offset_conv2(L3_offset))
        L3_fea = self.lrelu(self.L3_dcnpack(nbr_fea_l[2], L3_offset))
        # L2
        L2_offset = torch.cat([nbr_fea_l[1], ref_fea_l[1]], dim=1)
        L2_offset = self.lrelu(self.L2_offset_conv1(L2_offset))
        L3_offset = F.interpolate(L3_offset, scale_factor=2, mode='bilinear', align_corners=False)
        L2_offset = self.lrelu(self.L2_offset_conv2(torch.cat([L2_offset, L3_offset * 2], dim=1)))
        L2_offset = self.lrelu(self.L2_offset_conv3(L2_offset))
        L2_fea = self.L2_dcnpack(nbr_fea_l[1], L2_offset)
        L3_fea = F.interpolate(L3_fea, scale_factor=2, mode='bilinear', align_corners=False)
        L2_fea = self.lrelu(self.L2_fea_conv(torch.cat([L2_fea, L3_fea], dim=1)))
        # L1
        L1_offset = torch.cat([nbr_fea_l[0], ref_fea_l[0]], dim=1)
        L1_offset = self.lrelu(self.L1_offset_conv1(L1_offset))
        L2_offset = F.interpolate(L2_offset, scale_factor=2, mode='bilinear', align_corners=False)
        L1_offset = self.lrelu(self.L1_offset_conv2(torch.cat([L1_offset, L2_offset * 2], dim=1)))
        L1_offset = self.lrelu(self.L1_offset_conv3(L1_offset))
        L1_fea = self.L1_dcnpack(nbr_fea_l[0], L1_offset)
        L2_fea = F.interpolate(L2_fea, scale_factor=2, mode='bilinear', align_corners=False)
        L1_fea = self.L1_fea_conv(torch.cat([L1_fea, L2_fea], dim=1))
        # Cascading
        offset = torch.cat([L1_fea, ref_fea_l[0]], dim=1)
        offset = self.lrelu(self.cas_offset_conv1(offset))
        offset = self.lrelu(self.cas_offset_conv2(offset))
        L1_fea = self.lrelu(self.cas_dcnpack(L1_fea, offset))

        return L1_fea`

2. We use 3 DRDBs, the setting is: DRDB(6, 5, 64, 32).
3. Could you tell me more about this question? I'm not quite sure what you mean.

I will upload our codes asap. (Probably in early August).

**Comment by mtntruong** [comment]
Hello,

Thank you very much for your time.

For the last question, the part "Fusion Subnet for HDR Reconstruction" in [the manuscript](https://arxiv.org/abs/2105.10697) says "f^t_2 and f^l_2 denote the same reference features". I am still unclear about this sentence. As f^t_2 is the feature extracted from the middle LDR image by a convolutional layer, is f^l_2 = f^t_2? If not, where does f^l_2 come from?

I am very grateful for your help. Thank you.

**Comment by Pea-Shooter** [comment]
Yes, f^l_2 = f^t_2. We first extract the features through several conv layers and then feed them into the two branches.  f^l_2 and f^t_2 mean the same reference features.

**Comment by liuzhen03** [comment]
f^l_2 and f^t_2 generate from the same reference image, l and t mean linear domain and tonemapped domain respectively. Sorry for making a confusing statement in our paper.
